Modules/posixmodule.c

```diff
@@ -7651,6 +7651,115 @@ os_sched_getaffinity_impl(PyObject *module, pid_t pid)
 
 #endif /* HAVE_SCHED_H */
 
+#ifdef HAVE_POSIX_OPENPT
+/*[clinic input]
+os.posix_openpt -> int
+
+    oflag: int
+    /
+
+Open and return a file descriptor for a master pseudo-terminal device.
+
+Performs a posix_openpt() system call. The oflag argument is used to set
+file status flags and file access modes as specified in the manual page
+of posix_openpt() of your system.
+[clinic start generated code]*/
+
+static int
+os_posix_openpt_impl(PyObject *module, int oflag)
+/*[clinic end generated code: output=ee0bc2624305fc79 input=012468535878d78e]*/
+{
+    int fd;
+
+    fd = posix_openpt(oflag);
+    if (fd == -1)
+        posix_error();
+
+    return fd;
+}
+#endif /* HAVE_POSIX_OPENPT */
+
+#ifdef HAVE_GRANTPT
+/*[clinic input]
+os.grantpt
+
+    fd: fildes
+        File descriptor of a master pseudo-terminal device.
+    /
+
+Grant access to the slave pseudo-terminal device.
+
+Performs a grantpt() system call.
+[clinic start generated code]*/
+
+static PyObject *
+os_grantpt_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=dfd580015cf548ab input=7516c1a3cc2b9cfc]*/
+{
+    int ret;
+    PyOS_sighandler_t sig_saved;
+
+    sig_saved = PyOS_setsig(SIGCHLD, SIG_DFL);
```

Comment:
The POSIX `grantpt()` manpage (https://pubs.opengroup.org/onlinepubs/9699919799/) says

> The behavior of the grantpt() function is unspecified if the application has installed a signal handler to catch SIGCHLD signals.

What we are doing here is exactly how the fallback implementation of `os.openpty()` makes sure that behavior is not unpredictable: a `PyOS_setsig` call replaces the custom `SIGCHLD` handler with the default handler followed by the `grantpt()` call followed by restoring the original `SIGCHLD` handler.